### DISCUSSION TITLE
Custom checkbox bug / question, cannot trigger onChange event of the custom checkbox

### DISCUSSION THREAD

**Discussion by klendi** [react-hook-form/react-hook-form#4451] [Q&A] [answered]
**Describe the bug**
When using a custom checbox, and you forward ref from a component you can't use setValue to make a checkbox active. Or how to trigger the onChange event of the checkbox when setting the value with react-hook-form. Perhaps it isn't supposed to call the onChange event and I may be wrong.

**To Reproduce**
1. Create custom checkbox component.
2. Forward Ref from the component.
3. Set value using `setValue("name", true)`

**Codesandbox link (Required)**
https://codesandbox.io/s/react-hook-form-custom-checkbox-forked-h2u1i?file=/src/App.js

**Expected behavior**
To trigger onChange when you setValue

**Desktop (please complete the following information):**

- OS: Windows 10
- Browser Chrome

**Additional context**
Is there any way to trigger the event. Because the custom checkbox isn't changed at all. When you click on it it works. But with set value can't find a way on how to do it using react hook form.

**Comment by bluebill1049** [comment]
If you want to invoke `onChange`, use `useController`. setValue with register will directly update the dom value.

**Comment by klendi** [comment]
How does setValue of useForm works? How it doesn't trigger onChange, because I'm trying to keep a standard, using useForm instead of useControl. Is that possible ?

**Comment by klendi** [comment]
What I am having a problem with kinda "defaultValue", when I fetch the data I set the value, but with a custom checkbox that doesn't work. I don't have a way how to check the custom checkbox, I thought react hook form would do that and onChange event would be triggered so I would have the opportunity to change it. Without doing any hacks

**Comment by bluebill1049** [comment]
> when I fetch the data I set the value, but with a custom checkbox that doesn't work. 

share an example?

**Comment by klendi** [comment, marked as answer]
```  
const { register, handleSubmit, errors, setValue } = useForm();
  useEffect(() => {
    if (data) {
      setValue("username", data.username);
    }
  }, [data]);

return (
  <form onSubmit={handleSubmit(onSubmit)}>
      <Row>
        <Col md={5}>
          <Input
            placeholder="Username"
            label="Username"
            required
            name="username"
            ref={register({ required: true, minLength: 3 })}
          />
)
```
I even tried useController but that wasn't any of use. **I want to get when the react form hook changes the checkbox value event**. 
Couldn't find a way to do that, it wasn't working with react hook form, and I didn't want to make it controlled, so I made a method and exported inside the ref, using imperative handles
```
import React, { forwardRef, useImperativeHandle, useState } from "react";

interface CheckboxTypes {
  name?: string;
  label?: string;
  defaultChecked?: boolean;
}

const Checkbox = forwardRef(
  ({ label, name, defaultChecked }: CheckboxTypes, ref) => {
    const [checkboxState, setCheckboxState] = useState(false);

    const onChange = () => {
      debugger;
      setCheckboxState(!checkboxState);
    };

    useImperativeHandle(ref, () => ({
      set: (val: boolean) => {
        setCheckboxState(val);
      },
    }));

    return (
      <label>
        <CheckboxContainer>
          <HiddenCheckbox
            defaultChecked={defaultChecked}
            name={name}
            onChange={onChange}
            ref={ref}
          />
          <StyledCheckbox checked={checkboxState}>
            <Icon viewBox="0 0 24 24">
              <polyline points="20 6 9 17 4 12" />
            </Icon>
          </StyledCheckbox>
        </CheckboxContainer>
        <span className="ml-1">{label}</span>
      </label>
    );
  }
);

export { Checkbox };
```

and this how I set the values now

```
const setDefaultValues = (fields: InputField[], data: any, setValue) => {
  if (data) {
    fields.forEach(field => {
      if (field.type === "checkbox" && field.ref?.current) {
        field.ref?.current.set(data[field.name]);
      }

      setValue(field.name, data[field.name]);
    });
  }
};
```
This sets both the hidden checkbox value, and the styled div that is like a checkbox, so I can get it on getValues and also have it appear on the UI. This is my solution. using setValue from react-hook-form didn't trigger any event so there wasn't any way to do it. I wonder how other people have done it

**Reply by bluebill1049** [reply]
`setValue` will directly update the dom, and only watch/useWatch can be subscribed for the change. when setValue gets invoked for controller, we don't invoke `onChange` but update its value: 

https://codesandbox.io/s/angry-mclean-9018s?file=/src/App.js

**Reply by klendi** [reply]
I understand. Didn't know that, I thought it makes sense that it changes the dom, meaning it changes it, and that is the logic of onChange to get triggered on any change. Anyway thank you
